File: functional_effects/algebraic/result.py
"""Result[T, E] type for explicit error handling.

This module provides a Result type that represents either success (Ok) or failure (Err).
It eliminates the need for exceptions in business logic and makes error handling explicit.
"""

from collections.abc import Callable
from dataclasses import dataclass
from typing import Never


@dataclass(frozen=True)
class Ok[T]:
    """Success variant of Result containing a value."""

    value: T

    def is_ok(self) -> bool:
        """Check if this is an Ok result."""
        return True

    def is_err(self) -> bool:
        """Check if this is an Err result."""
        return False

    def map[U](self, f: Callable[[T], U]) -> Ok[U]:
        """Map the success value with function f."""
        return Ok(f(self.value))

    def flat_map[U, E](self, f: Callable[[T], "Result[U, E]"]) -> "Result[U, E]":
        """Chain another Result-producing computation (monadic bind)."""
        return f(self.value)

    def map_err[E, F](self, _f: Callable[[E], F]) -> Ok[T]:
        """Map the error value (no-op for Ok)."""
        return self

    def unwrap(self) -> T:
        """Extract the value (safe for Ok)."""
        return self.value

    def unwrap_or[U](self, _default: U) -> T:
        """Return the value or default (returns value for Ok)."""
        return self.value

    def unwrap_err(self) -> Never:
        """Extract the error (raises for Ok)."""
        raise ValueError(f"Called unwrap_err on Ok: {self.value}")


@dataclass(frozen=True)
class Err[E]:
    """Failure variant of Result containing an error."""

    error: E

    def is_ok(self) -> bool:
        """Check if this is an Ok result."""
        return False

    def is_err(self) -> bool:
        """Check if this is an Err result."""
        return True

    def map[T, U](self, _f: Callable[[T], U]) -> Err[E]:
        """Map the success value (no-op for Err)."""
        return self

    def flat_map[T, U](self, _f: Callable[[T], "Result[U, E]"]) -> Err[E]:
        """Chain another Result-producing computation (no-op for Err)."""
        return self

    def map_err[F](self, f: Callable[[E], F]) -> Err[F]:
        """Map the error value with function f."""
        return Err(f(self.error))

    def unwrap(self) -> Never:
        """Extract the value (raises for Err)."""
        raise ValueError(f"Called unwrap on Err: {self.error}")

    def unwrap_or[U](self, default: U) -> U:
        """Return the value or default (returns default for Err)."""
        return default

    def unwrap_err(self) -> E:
        """Extract the error (safe for Err)."""
        return self.error


# Type alias for the Result union using PEP 695 syntax
type Result[T, E] = Ok[T] | Err[E]


def fold_result[
    T, E, U
](result: Result[T, E], on_ok: Callable[[T], U], on_err: Callable[[E], U],) -> U:
    """Fold pattern for exhaustive Result handling.

    Args:
        result: The Result to fold over
        on_ok: Function to apply if result is Ok
        on_err: Function to apply if result is Err

    Returns:
        The result of applying the appropriate function

    Example:
        >>> result = Ok(42)
        >>> fold_result(result, lambda x: x * 2, lambda e: 0)
        84
    """
    match result:
        case Ok(value=v):
            return on_ok(v)
        case Err(error=e):
            return on_err(e)
        case _:  # pragma: no cover
            # This should never happen due to Result type constraint
            raise AssertionError(f"Invalid Result variant: {result}")

File: demo/healthhub/frontend/src/algebraic/Result.ts
/**
 * Result<T, E> type for functional error handling.
 *
 * A Result represents either success (Ok) or failure (Err),
 * making error handling explicit and eliminating exceptions.
 */

export interface Ok<T> {
  readonly type: 'Ok'
  readonly value: T
}

export interface Err<E> {
  readonly type: 'Err'
  readonly error: E
}

export type Result<T, E> = Ok<T> | Err<E>

// Constructors
export const ok = <T, E = never>(value: T): Result<T, E> => ({
  type: 'Ok',
  value,
})

export const err = <T = never, E = never>(error: E): Result<T, E> => ({
  type: 'Err',
  error,
})

// Type guards
export const isOk = <T, E>(result: Result<T, E>): result is Ok<T> =>
  result.type === 'Ok'

export const isErr = <T, E>(result: Result<T, E>): result is Err<E> =>
  result.type === 'Err'

// Core operations
export const map = <T, U, E>(
  f: (value: T) => U
) => (result: Result<T, E>): Result<U, E> =>
  isOk(result) ? ok(f(result.value)) : result

export const mapErr = <T, E, F>(
  f: (error: E) => F
) => (result: Result<T, E>): Result<T, F> =>
  isErr(result) ? err(f(result.error)) : result

export const flatMap = <T, U, E>(
  f: (value: T) => Result<U, E>
) => (result: Result<T, E>): Result<U, E> =>
  isOk(result) ? f(result.value) : result

export const fold = <T, E, U>(
  onOk: (value: T) => U,
  onErr: (error: E) => U
) => (result: Result<T, E>): U =>
  isOk(result) ? onOk(result.value) : onErr(result.error)

// Utility functions
export const unwrapOr = <T, E>(defaultValue: T) => (result: Result<T, E>): T =>
  isOk(result) ? result.value : defaultValue

export const unwrapOrElse = <T, E>(
  f: (error: E) => T
) => (result: Result<T, E>): T =>
  isOk(result) ? result.value : f(result.error)

// Convert from Promise
export const fromPromise = async <T>(
  promise: Promise<T>
): Promise<Result<T, Error>> => {
  try {
    const value = await promise
    return ok(value)
  } catch (error) {
    return err(error instanceof Error ? error : new Error(String(error)))
  }
}

// Combine multiple Results
export const all = <T, E>(results: Result<T, E>[]): Result<T[], E> => {
  const values: T[] = []
  for (const result of results) {
    if (isErr(result)) return result
    values.push(result.value)
  }
  return ok(values)
}
